File: package/android/src/main/java/com/margelo/nitro/realfetch/HybridOutputStream.kt
package com.margelo.nitro.realfetch

import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import java.io.OutputStream

class HybridOutputStream(private val stream: OutputStream) : HybridOutputStreamSpec() {
  override fun write(buffer: ArrayBuffer): Promise<Unit> {
    val byteBuffer = buffer.getBuffer(false)
    val bytes = ByteArray(buffer.size)
    byteBuffer.get(bytes)

    return Promise.async {
      stream.write(bytes)
    }
  }

  override fun open() {
  }

  override fun close() {
    stream.flush()
    stream.close()
  }

  override val memorySize: Long
    get() = 0L
}

File: package/ios/HybridOutputStream.swift
import Foundation
import NitroModules

class HybridOutputStream: HybridOutputStreamSpec {
  let stream: OutputStream

  init(stream: OutputStream) {
    self.stream = stream
  }

  func open() throws {
    stream.open()
  }

  func write(buffer: ArrayBuffer) throws -> Promise<Void> {
    let promise = Promise<Void>()

    let data = buffer.data
    let length = buffer.size

    Task {
      let bytesWritten = stream.write(data, maxLength: length)
      if bytesWritten == length {
        promise.resolve(withResult: ())
      } else {
        promise.reject(
          withError: stream.streamError
            ?? RuntimeError.error(withMessage: "Unexpected error writing to stream"))
      }
    }

    return promise
  }

  func close() {
    stream.close()
  }

  var memorySize: Int {
    return 0
  }
}
